Add bookings

Vexvl opened the pull request.

Vexvl committed:
added user JPA

Vexvl committed:
feat : AbsenceException.java
       AccessDeniedException.java
       Booking.java
       BookingController.java
       BookingDto.java
       BookingMapper.java
       BookingRepository.java
       BookingService.java
       BookingServiceImpl.java
       BookingState.java
       BookingStatus.java
       Comment.java
       CommentController.java
       CommentDto.java
       CommentMapper.java
       CommentRepository.java
       CommentService.java
       CommentServiceImpl.java
       EmailDuplicateException.java
       ExceptionsHandler.java
       InvalidRequestException.java
       InvalidStatusException.java
       Item.java
       ItemController.java
       ItemDto.java
       ItemDtoBookingComments.java
       ItemMapper.java
       ItemRepository.java
       ItemRequest.java
       ItemRequestController.java
       ItemRequestDto.java
       ItemService.java
       ItemServiceImpl.java
       ItemUnavailableException.java
       NoBookingException.java
       NoPossessionException.java
       OwnerBookingException.java
       schema.sql
       ShareItApp.java
       UnknownStatusException.java
       User.java
       UserController.java
       UserService.java

Vexvl committed:
Delete db directory

Vexvl committed:
feat : AbsenceException.java
       AccessDeniedException.java
       Booking.java
       BookingController.java
       BookingDto.java
       BookingMapper.java
       BookingRepository.java
       BookingService.java
       BookingServiceImpl.java
       BookingState.java
       BookingStatus.java
       Comment.java
       CommentController.java
       CommentDto.java
       CommentMapper.java
       CommentRepository.java
       CommentService.java
       CommentServiceImpl.java
       EmailDuplicateException.java
       ExceptionsHandler.java
       InvalidRequestException.java
       InvalidStatusException.java
       Item.java
       ItemController.java
       ItemDto.java
       ItemDtoBookingComments.java
       ItemMapper.java
       ItemRepository.java
       ItemRequest.java
       ItemRequestController.java
       ItemRequestDto.java
       ItemService.java
       ItemServiceImpl.java
       ItemUnavailableException.java
       NoBookingException.java
       NoPossessionException.java
       OwnerBookingException.java
       schema.sql
       ShareItApp.java
       UnknownStatusException.java
       User.java
       UserController.java
       UserService.java

Vexvl committed:
style changes

Vexvl committed:
style changes

Vexvl committed:
style changes

Vexvl committed:
style changes

Vexvl committed:
style changes

Vexvl committed:
style changes

Vexvl committed:
style changes

sergeymamonov commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -1,12 +1,46 @@
+    }
+
+    @PatchMapping("/{bookingId}")
+    public BookingDto changeBookingStatus(@RequestHeader("X-Sharer-User-Id") Long ownerId, @PathVariable Long bookingId, @RequestParam String approved) {
```
Старайтесь не выходить на длину строки в 120 символов.

sergeymamonov commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,161 @@
+            case "FALSE":
+                booking.setStatus(BookingStatus.REJECTED);
+                break;
+        }
```
Лучше всегда делать ветку default на случай, если мы получим статус, который нам неизвестен, чтобы можно было кинуть ошибку или что-то еще сделать в такой ситуации. Проверьте везде, где используете switch.

sergeymamonov replied:
А еще лучше сделать approved логическим типом данных.

sergeymamonov commented on src/main/java/ru/practicum/shareit/handler/ExceptionsHandler.java:
```diff
@@ -0,0 +1,84 @@
+    @ExceptionHandler(AbsenceException.class)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public Map<String, String> handleAbsenceException(AbsenceException e) {
+        return Map.of("error", "Такого элемента нет", "errorMessage", "Ошибка на сервере");
```
Ошибка на сервере везде выглядит немного странно. Тем более, что при получении невалидных данных никакой ошибки на сервере, в том понимании, что произошла какая-то непредвиденная ошибка, нет.

sergeymamonov commented on src/main/java/ru/practicum/shareit/handler/ExceptionsHandler.java:
```diff
@@ -0,0 +1,84 @@
+
+    @ExceptionHandler(EmailDuplicateException.class)
+    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
+    public Map<String, String> handleEmailDuplicateException(EmailDuplicateException e) {
```
По названию ошибки это скорее 409 статус.

sergeymamonov commented on src/main/java/ru/practicum/shareit/handler/ExceptionsHandler.java:
```diff
@@ -0,0 +1,84 @@
+
+    @ExceptionHandler(UnknownStatusException.class)
+    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
+    public Map<String, String> handleUnknownStatusException(UnknownStatusException e) {
```
Скорее 400 статус.

sergeymamonov replied:
Тут либо статус изменить, либо название ошибки, потому что неизвестный статус не должен вызывать 500 ответ. 500 ответ - это ошибка в коде.

sergeymamonov commented on src/main/java/ru/practicum/shareit/item/service/impl/ItemServiceImpl.java:
```diff
@@ -1,56 +1,127 @@
+        for (Item item : itemList) {
+            ItemDtoBookingComments itemDto = itemMapper.toItemDtoBookingComments(item, commentRepository.findAllByItem(item));
+
+            List<Booking> nextBookingsList = bookingRepository.findNextOrderedBookingsByItemId(itemDto.getId());
```
Не очень хорошо делать запросы к БД в цикле. Это большая нагрузка на БД.

sergeymamonov commented on src/main/java/ru/practicum/shareit/item/service/impl/ItemServiceImpl.java:
```diff
@@ -1,56 +1,127 @@
+
+            itemDtoList.add(itemDto);
+        }
+        ;
```
Лишняя точка с запятой.

sergeymamonov commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,55 @@
+    end_date   TIMESTAMP WITHOUT TIME ZONE             NOT NULL,
+    item_id    BIGINT                                  NOT NULL,
+    booker_id  BIGINT                                  NOT NULL,
+    status     VARCHAR                                 NOT NULL,
```
Размер varchar должен быть ограничен.

Vexvl replied:
Спасибо за замечания, всё исправил

Vexvl committed:
fix: getItemsByOwner,
style changes: Impl
fix + style changes: ExceptionsHandler
fix: restraints schema

Vexvl committed:
fix: getItemsByOwner,
style changes: Impl
fix + style changes: ExceptionsHandler
fix: restraints schema

Vexvl committed:
fix: getItemsByOwner,
style changes: Impl
fix + style changes: ExceptionsHandler
fix: restraints schema

Vexvl committed:
fix: getItemsByOwner,
style changes: Impl
fix + style changes: ExceptionsHandler
fix: restraints schema

sergeymamonov commented on src/main/java/ru/practicum/shareit/item/service/impl/ItemServiceImpl.java:
```diff
@@ -1,56 +1,147 @@
+
+        for (Item item : itemList) {
+            ItemDtoBookingComments itemDto = itemMapper.toItemDtoBookingComments(item,
+                    commentRepository.findAllByItem(item));
```
Запрос к БД в цикле.

sergeymamonov reviewed (approved):
Устранить оставшиеся замечания.
